src/foam/util/AddressUtil.java

```diff
@@ -69,48 +93,89 @@ public static String[] parseAddress(String address1, String address2) throws Ill
     return new String[] { streetNumber, streetName };
   }
 
-  public static String normalizeRegion(X x, String country, String regionCode) {
-    if ( SafetyUtil.isEmpty(regionCode) || SafetyUtil.isEmpty(country) ) {
-      return regionCode;
+  // TODO: country defaults to CA if not given. Remove when we can.
+  public static String normalizeRegion(X x, String country, String region) {
+    if ( SafetyUtil.isEmpty(region) ) return region;
+
+    if ( SafetyUtil.isEmpty(country) ) {
+      country = "CA";
+    }
+
+    country = country.trim();
+    region = region.trim();
+
+    Region regionObject = lookupRegion(x, country, region);
+    if ( regionObject != null ) {
+      return regionObject.getCode();
+    } else {
+      return country + "-" + region;
+    }
+  }
+
+  // TODO: country defaults to CA if not given. Remove when we can.
+  public static String normalizeRegion(X x, String country, String region, String postalCode) {
+    if ( SafetyUtil.isEmpty(country) ) {
+      country = "CA";
+    }
+
+    country = country.trim();
+    if ( ! SafetyUtil.isEmpty(region) ) region = region.trim();
+    if ( ! SafetyUtil.isEmpty(postalCode) ) postalCode = postalCode.trim();
+
+    if ( ! SafetyUtil.isEmpty(region) ) {
+      Region regionObject = lookupRegion(x, country, region);
+      if ( regionObject != null ) return regionObject.getCode();
     }
-    
-    String normalizedRegion = country + "-" + regionCode;
 
-    DAO regionDAO = (DAO) x.get("regionDAO");
-    Region region = (Region) regionDAO.find(AND(
+    if ( ! SafetyUtil.isEmpty(postalCode) ) {
+      String regionCode = lookupRegionCodeWithPostalCode(x, country, postalCode);
+      if ( ! SafetyUtil.isEmpty(regionCode) ) return regionCode; 
+    }
+
+    return SafetyUtil.isEmpty(region) ? region : country + "-" + region;
+  }
+
+  private static Region lookupRegion(X x, String country, String region) {
+    return (Region) ((DAO) x.get("regionDAO")).find(AND(
       EQ(Region.COUNTRY_ID, country),
       OR(
-        EQ(Region.ISO_CODE, regionCode),
-        EQ(Region.CODE, regionCode),
-        IN(regionCode.toUpperCase(), Region.ALTERNATIVE_NAMES)
+        EQ(Region.ISO_CODE, region),
+        EQ(Region.CODE, region),
+        IN(region.toUpperCase(), Region.ALTERNATIVE_NAMES)
       )
     ));
-
-    if ( region != null ) {
-      normalizedRegion = region.getCode();
+  }
+  
+  // TODO: lookup logic defaults to CA logic if not given. Remove when we can.
+  private static String lookupRegionCodeWithPostalCode(X x, String country, String postalCode) {
+    if ( SafetyUtil.isEmpty(postalCode) ) return "";
+
+    switch(country) {
+      case "CA":
+      default: 
```

Comment:
@adamfox-nanopay should we only use ca logic when country is CA or empty? If country is set to something other than CA whose postal code starts with a letter, the method breaks.

Reply:
if country is empty you are now setting it to CA so we can remove default here as it will be CA for CA and empty